File: org.scala-ide.sdt.aspects/src/scala/tools/eclipse/contribution/weaving/jdt/util/ReflectionUtils.java
/*
 * Copyright 2005-2009 LAMP/EPFL
 */
// $Id$

package scala.tools.eclipse.contribution.weaving.jdt.util;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.security.AccessController;
import java.security.PrivilegedActionException;
import java.security.PrivilegedExceptionAction;

public class ReflectionUtils {
  public static <T> Constructor<T> getConstructor(final Class<T> clazz, final Class<?> ... paramTypes) {
    try {
      return AccessController.doPrivileged(new PrivilegedExceptionAction<Constructor<T>>() {
        public Constructor<T> run() throws Exception {
          Constructor<T> ctor = clazz.getDeclaredConstructor(paramTypes);
          ctor.setAccessible(true);
          return ctor;
        }
      });
    }
    catch(PrivilegedActionException ex) {
      throw new RuntimeException(ex);
    }
  }

  public static <T> Method getMethod(final Class<T> clazz, final String name, final Class<?> ... paramTypes) {
    try {
      return AccessController.doPrivileged(new PrivilegedExceptionAction<Method>() {
        public Method run() throws Exception {
          Method m = clazz.getDeclaredMethod(name, paramTypes);
          m.setAccessible(true);
          return m;
        }
      });
    }
    catch(PrivilegedActionException ex) {
      throw new RuntimeException(ex);
    }
  }

  public static <T> Field getField(final Class<T> clazz, final String name) {
    try {
      return AccessController.doPrivileged(new PrivilegedExceptionAction<Field>() {
        public Field run() throws Exception {
          Field f = clazz.getDeclaredField(name);
          f.setAccessible(true);
          return f;
        }
      });
    }
    catch(PrivilegedActionException ex) {
      throw new RuntimeException(ex);
    }
  }
}

File: org.scala-ide.sdt.core/src/scala/tools/eclipse/util/ReflectionUtils.scala
/*
 * Copyright 2005-2010 LAMP/EPFL
 */
// $Id$

package scala.tools.eclipse.util

import java.lang.reflect.{ AccessibleObject, Constructor }
import java.security.{ AccessController, PrivilegedAction, PrivilegedActionException }

trait ReflectionUtils {
  def getConstructor[T](clazz : Class[T], paramTypes : Class[_]*) =
    privileged {
      val ctor = clazz.getConstructor(paramTypes : _*)
      ctor.setAccessible(true)
      ctor
    }

  def getMethod[T](clazz : Class[T], name : String, paramTypes : Class[_]*) =
    privileged {
      val method = clazz.getMethod(name, paramTypes : _*)
      method.setAccessible(true)
      method
    }

  def getField[T](clazz : Class[T], name : String) =
    privileged {
      val field = clazz.getField(name)
      field.setAccessible(true)
      field
    }

  def getDeclaredConstructor[T](clazz : Class[T], paramTypes : Class[_]*) =
    privileged {
      val ctor = clazz.getDeclaredConstructor(paramTypes : _*)
      ctor.setAccessible(true)
      ctor
    }

  def getDeclaredMethod[T](clazz : Class[T], name : String, paramTypes : Class[_]*) =
    privileged {
      val method = clazz.getDeclaredMethod(name, paramTypes : _*)
      method.setAccessible(true)
      method
    }

  def getDeclaredField[T](clazz : Class[T], name : String) =
    privileged {
      val field = clazz.getDeclaredField(name)
      field.setAccessible(true)
      field
    }

  def privileged[T](body : => T) = {
    try {
      AccessController.doPrivileged(new PrivilegedAction[T] {
        def run() = { body }
      })
    }
    catch {
      case ex : PrivilegedActionException => throw ex.getCause
    }
  }
}
